src/codec.rs

```diff
@@ -286,6 +284,52 @@ pub trait DecodeLength {
 	fn len(self_encoded: &[u8]) -> Result<usize, Error>;
 }
 
+/// A decoding context.
+///
+/// Used in [`Decode::decode_into`] to allow the implementation to explicitly
+/// assert that the `MaybeUninit` passed into that function was properly initialized.
+pub struct DecodeContext(PhantomData<*const ()>);
```

Comment:
Is there more "value" in this than described in the docs? If no, can we not just add a `unsafe fn new` to `DecodeFinished`?

Reply:
Hmm.... the idea was that it's only possible to materialize a `DecodeFinished` inside of the `decode_into` (because that's the only place where `DecodeContext` is available) but perhaps you're right. Since the function to create a `DecodeFinished` is `unsafe` anyway the extra `DecodeContext` doesn't really bring too much benefit.

I'll simplify it.